ResearchKit/Common/ORKSurveyAnswerCellForText.m

```diff
@@ -263,61 +285,36 @@ - (void)answerDidChange {
 
 #pragma mark - UITextFieldDelegate
 
-- (BOOL)correctValueIfNeeded {
-    ORKAnswerFormat *impliedFormat = [self.step impliedAnswerFormat];
-    NSAssert([impliedFormat isKindOfClass:[ORKTextAnswerFormat class]], @"answerFormat should be ORKTextAnswerFormat type instance.");
-    NSString *text = self.textField.text;
-    NSInteger maxLength = [(ORKTextAnswerFormat *)impliedFormat maximumLength];
-    if (maxLength > 0 && [text length] > maxLength) {
-        NSString *corrected = [text substringToIndex:maxLength];
-        ORK_Log_Debug(@"%@ -> %@", text, corrected);
-        
-        self.textField.text = corrected;
-        return YES;
-    } else {
-        return NO;
-    }
-}
-
-- (BOOL)correctionNeededForText:(NSString *)text {
+- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
     ORKAnswerFormat *impliedFormat = [self.step impliedAnswerFormat];
     NSAssert([impliedFormat isKindOfClass:[ORKTextAnswerFormat class]], @"answerFormat should be ORKTextAnswerFormat type instance.");
-    NSInteger maxLength = [(ORKTextAnswerFormat *)impliedFormat maximumLength];
-    if (maxLength > 0 && [text length] > maxLength) {
-        return YES;
-    }
-    return NO;
-}
-
-- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
+    
     NSString *text = [textField.text stringByReplacingCharactersInRange:range withString:string];
-    if (! [self correctionNeededForText:text]) {
-        [self ork_setAnswer:[text length] ? text : ORKNullAnswerValue()];
-    } else {
+    
+    NSInteger maxLength = [(ORKTextAnswerFormat *)impliedFormat maximumLength];
+    
+    if(maxLength > 0 && [text length] > maxLength)
+    {
+        NSString *string = [NSString stringWithFormat:ORKLocalizedString(@"TEXT_ANSWER_ALERT_MESSAGE", nil), [@(maxLength) stringValue]];
+        [self showValidityAlertWithMessage:string];
         return NO;
     }
     
+    [self ork_setAnswer:[text length] ? text : ORKNullAnswerValue()];
+    
     return YES;
 }
```

Comment:
The shared logic in `textField:shouldChangeCharactersInRange:replacementString:` and `textView:shouldChangeTextInRange:replacementText:` ought to consolidate into one piece. 

And I feels like `[NSString stringWithFormat:ORKLocalizedString(@"TEXT_ANSWER_ALERT_MESSAGE", nil), [@(maxLength) stringValue]];` should be from `localizedInvalidValueStringWithAnswerString:`. Could let `localizedInvalidValueStringWithAnswerString:` has some logic to decide which message to show.